### PR TITLE
Fix utp connection leak on cancel

### PR DESCRIPTION
**Pull request by KonradStaniec**
Fixes https://github.com/status-im/nimbus-eth1/issues/1075 by bumping nim-eth and adding fixes to portal_protocol.

Summary: The main culprit of this leak is that cancellation and `await` do not play with each other very well when allocating some resource inside the `async` procs.

Consider simplified code:
```
import
  chronos

type SomeResourceWhichNeedToBeReleased = object

proc freeResource(res: SomeResourceWhichNeedToBeReleased): void = discard

proc use(res: SomeResourceWhichNeedToBeReleased): void= discard

proc allocateResource: Future[SomeResourceWhichNeedToBeReleased] = discard

proc someProcWhichUsesResource: Future[void] {.async.} =
  let resource = await allocateResource()
  try:
    resource.use()
  finally:
    resource.freeResource()

proc someProcWhichStartAndCancel() {.async.} =
  let fut = someProcWhichUsesResource()

  fut.cancel()

```

What was happening in current case was:
1. When calling `fut.cancel()`, the path with the [child](https://github.com/status-im/nim-chronos/blob/master/chronos/asyncfutures2.nim#L287) is taken where child future is `allocateResource` future (and the parent future is `someProcWhichUsesResource`)
2. But child future have already finished but not yet given control to the corresponding yield statement, so child cancel returns [false](https://github.com/status-im/nim-chronos/blob/master/chronos/asyncfutures2.nim#L284). So before cancel return, only the flag `mustCancel` on cancelled future is set to [true](https://github.com/status-im/nim-chronos/blob/master/chronos/asyncfutures2.nim#L295)
3. In the next eventLoop step control is given to the yield statement in `await allocateResource()` [yield](https://github.com/status-im/nim-chronos/blob/master/chronos/asyncmacro2.nim#L274), as the child future already finished (`allocateResource` future is child of `someProcWhichUsesResource` future ). As this child future already finished and not cancelled, resource was already allocated so to clean it up, call to `freeResource` is needed.
4. But as `mustCancel`  was set to true in point `2` for parent future, `await` actually throws and cancellation [exception](https://github.com/status-im/nim-chronos/blob/master/chronos/asyncmacro2.nim#L287) 
5. So it leads to the situation that there is no way to release resources allocated by `SomeResourceWhichNeedToBeReleased`

### CODE REVIEW

File: fluffy/network/wire/portal_protocol.nim [hunk 1]
```diff
@@ -614,37 +614,53 @@ proc findContent*(p: PortalProtocol, dst: Node, contentKey: ByteList):
...
+        # A FIN does not necessarily mean that the data read is complete. Further
+        # validation is required, using a length prefix here might be beneficial for
+        # this.
+        let readFut = socket.read()
+        if await readFut.withTimeout(p.stream.readTimeout):
+          let content = readFut.read
+          return ok(FoundContent(src: dst, kind: Content, content: content))
+        else :
+          return err("Reading data from socket timed out, content request failed")
+      finally:
```
Comment (kdeme) [inline fluffy/network/wire/portal_protocol.nim:651]: We have been avoiding the use of finally in async procs due to bugs on finally + closure iterators (and also issues when raising in finally)
I don't know the exact state of this now but considering following still opens bugs I'd stay away from it:
https://github.com/nim-lang/Nim/issues/15501
https://github.com/status-im/nim-chronos/issues/132
https://github.com/nim-lang/Nim/issues/11887
(You can find more if you search for finally / defer )

Comment (KonradStaniec) [reply fluffy/network/wire/portal_protocol.nim]: that nice to know 👍 rewritten it to catching cancelled exception and closing socket there

**Comment by kdeme** [comment]
> What was happening in current case was:

I haven't dived deep into this but would you say than that setting the `mustCancel` flag to true in the case that the child cancel failed is a bug?

**Comment by KonradStaniec** [comment]
tbh at first glance it smells a little bit like bug, as it the child future already finished then we should definitely give back control to parent future instead of just throwing (and possibly leaving some dangling resources).

That saying, to be sure I also would need to analyse how futures interplay with cancellation a bit more to be sure, especially as `mustCancel` was introduced to fix some kinda race: https://github.com/status-im/nim-chronos/pull/107

**Pull request merged**